src/Microsoft.AspNet.OData.Shared/Builder/DerivedTypeConstraintSet.cs

```diff
@@ -0,0 +1,72 @@
+﻿// Copyright (c) Microsoft Corporation.  All rights reserved.
+// Licensed under the MIT License.  See License.txt in the project root for license information.
+
+using Microsoft.AspNet.OData.Common;
+using Microsoft.OData.UriParser;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Microsoft.AspNet.OData.Builder
+{
+    internal class DerivedTypeConstraintSet : HashSet<Type>
+    {
+        /// <summary>
+        /// 
+        /// </summary>
+        public DerivedTypeConstraintSet()
+        {
+        }
+
+        public DerivedTypeConstraintSet(Type baseType)
+        {
+            ClrBaseType = baseType;
+        }
+
+        public Type ClrBaseType { get; set; }
+
+        public void ValidateAndAddConstraints(IEnumerable<Type> types)
+        {
+            if (ClrBaseType == null)
+            {
+                Error.InvalidOperation(SRResources.BaseTypeNotSpecified);
+            }
+
+            ValidateAndAddConstraints(ClrBaseType, types);
+        }
+
+        public void ValidateAndAddConstraints(Type baseType, IEnumerable<Type> types)
```

Comment:
>baseType [](start = 51, length = 8)

This dual-capability of having a ClrBaseType declared on the DerivedTypeConstraintSet *and* having overloads that allow passing the base type is a bit strange.  It took me a bit to figure out that what (I think) you're doing is accounting for the fact that the base type may not be known when the DerivedTypeConstraintSet is created in the element constructor.  But it's still strange, as the baseType passed in to ValidateAndAddConstraints could have nothing to do with the type of the element on which the DerivedTypeConstraint is defined.

What if, instead, the element lazily created the DerivedTypeConstraintSet the first time it was accessed, and returned an error if the type had not yet been set of the element.  Then you wouldn't ever need to pass in the baseType, or worry if the passed in baseType was at all related to the ClrBaseType (if specified).  You could then replace ValidateAndAddSingleConstraint with an override of the Add method that did the validation before calling base.Add. This would make it easier to expose as a public property on the element.

Reply:
I tried to address many things. I still have the validation in place though because I felt that invalidating the model later will make it harder for developers to find the issue. We can discuss if the current design meets the need. 

---
In reply to: [384819539](https://github.com/OData/WebApi/pull/1977#discussion_r384819539) [](ancestors = 384819539)